File: contests/Aoc/2025/6/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	part2 := flag.Bool("part2", false, "enable part 2 interpretation (columns right-to-left)")
	flag.Parse()

	lines, width, err := readLines()
	if err != nil {
		fmt.Fprintln(os.Stderr, "read error:", err)
		os.Exit(1)
	}

	if len(lines) == 0 {
		fmt.Println(0)
		return
	}

	grid := padLines(lines, width)
	total, err := solveGrid(grid, *part2)
	if err != nil {
		fmt.Fprintln(os.Stderr, "solve error:", err)
		os.Exit(1)
	}

	fmt.Println(total)
}

func readLines() ([]string, int, error) {
	scanner := bufio.NewScanner(os.Stdin)
	lines := make([]string, 0)
	maxLen := 0

	for scanner.Scan() {
		line := scanner.Text()
		if len(line) > maxLen {
			maxLen = len(line)
		}
		lines = append(lines, line)
	}

	if err := scanner.Err(); err != nil {
		return nil, 0, err
	}

	return lines, maxLen, nil
}

func padLines(lines []string, width int) [][]byte {
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		row := []byte(line)
		if len(row) < width {
			row = append(row, make([]byte, width-len(row))...)
		}
		grid[i] = row
	}
	return grid
}

func solveGrid(grid [][]byte, part2 bool) (int, error) {
	height := len(grid)
	if height == 0 {
		return 0, nil
	}

	width := len(grid[0])
	opRow := height - 1

	separator := make([]bool, width)
	hasSeparator := false
	for col := 0; col < width; col++ {
		allSpace := true
		for row := 0; row < height; row++ {
			if grid[row][col] != ' ' {
				allSpace = false
				break
			}
		}
		separator[col] = allSpace
		if allSpace {
			hasSeparator = true
		}
	}

	type problem struct {
		op      byte
		col     int
		numbers []int
	}

	ops := make([]problem, 0)
	for col, ch := range grid[opRow] {
		if ch == '+' || ch == '*' {
			ops = append(ops, problem{op: ch, col: col, numbers: make([]int, 0)})
		}
	}
	if len(ops) == 0 {
		return 0, fmt.Errorf("no operators found")
	}

	nearestOp := func(col int) int {
		bestIdx := 0
		bestDist := abs(col - ops[0].col)
		for i := 1; i < len(ops); i++ {
			dist := abs(col - ops[i].col)
			if dist < bestDist || (dist == bestDist && ops[i].col > ops[bestIdx].col) {
				bestDist = dist
				bestIdx = i
			}
		}
		return bestIdx
	}

	if hasSeparator {
		total := 0
		for col := 0; col < width; {
			for col < width && separator[col] {
				col++
			}
			start := col
			for col < width && !separator[col] {
				col++
			}
			if start == col {
				continue
			}
			var value int
			var err error
			if part2 {
				value, err = evaluateSegmentPart2(grid, start, col)
			} else {
				value, err = evaluateSegment(grid, start, col)
			}
			if err != nil {
				return 0, err
			}
			total += value
		}
		return total, nil
	}

	if part2 {
		// Each column encodes a single number (top to bottom); assign to nearest operator.
		for col := 0; col < width; col++ {
			var b strings.Builder
			for row := 0; row < opRow; row++ {
				ch := grid[row][col]
				if ch >= '0' && ch <= '9' {
					b.WriteByte(ch)
				}
			}
			if b.Len() == 0 {
				continue
			}
			value, err := strconv.Atoi(b.String())
			if err != nil {
				return 0, fmt.Errorf("column %d: %w", col, err)
			}
			idx := nearestOp(col)
			ops[idx].numbers = append(ops[idx].numbers, value)
		}
	} else {
		for row := 0; row < opRow; row++ {
			line := grid[row]
			col := 0
			for col < width {
				if line[col] < '0' || line[col] > '9' {
					col++
					continue
				}
				start := col
				for col < width && line[col] >= '0' && line[col] <= '9' {
					col++
				}
				raw := string(line[start:col])
				value, err := strconv.Atoi(strings.TrimSpace(raw))
				if err != nil {
					return 0, fmt.Errorf("row %d columns %d-%d: %w", row, start, col, err)
				}
				center := (start + col - 1) / 2
				idx := nearestOp(center)
				ops[idx].numbers = append(ops[idx].numbers, value)
			}
		}
	}

	total := 0
	for _, p := range ops {
		if len(p.numbers) == 0 {
			return 0, fmt.Errorf("no numbers found for operator at column %d", p.col)
		}
		switch p.op {
		case '+':
			sum := 0
			for _, n := range p.numbers {
				sum += n
			}
			total += sum
		case '*':
			product := 1
			for _, n := range p.numbers {
				product *= n
			}
			total += product
		default:
			return 0, fmt.Errorf("unhandled operator %q at column %d", p.op, p.col)
		}
	}

	return total, nil
}

func evaluateSegment(grid [][]byte, start, end int) (int, error) {
	height := len(grid)
	opRow := height - 1

	op := byte(0)
	for _, ch := range grid[opRow][start:end] {
		if ch != ' ' {
			op = ch
			break
		}
	}
	if op != '+' && op != '*' {
		return 0, fmt.Errorf("invalid operator in columns %d-%d", start, end)
	}

	numbers := make([]int, 0, height-1)
	for row := 0; row < opRow; row++ {
		raw := strings.TrimSpace(string(grid[row][start:end]))
		if raw == "" {
			continue
		}
		value, err := strconv.Atoi(raw)
		if err != nil {
			return 0, fmt.Errorf("row %d columns %d-%d: %w", row, start, end, err)
		}
		numbers = append(numbers, value)
	}

	if len(numbers) == 0 {
		return 0, fmt.Errorf("no numbers found in columns %d-%d", start, end)
	}

	switch op {
	case '+':
		sum := 0
		for _, n := range numbers {
			sum += n
		}
		return sum, nil
	case '*':
		product := 1
		for _, n := range numbers {
			product *= n
		}
		return product, nil
	}

	return 0, fmt.Errorf("unhandled operator %q", op)
}

func evaluateSegmentPart2(grid [][]byte, start, end int) (int, error) {
	height := len(grid)
	opRow := height - 1

	op := byte(0)
	for _, ch := range grid[opRow][start:end] {
		if ch != ' ' {
			op = ch
			break
		}
	}
	if op != '+' && op != '*' {
		return 0, fmt.Errorf("invalid operator in columns %d-%d", start, end)
	}

	numbers := make([]int, 0)
	for col := end - 1; col >= start; col-- {
		var b strings.Builder
		for row := 0; row < opRow; row++ {
			ch := grid[row][col]
			if ch >= '0' && ch <= '9' {
				b.WriteByte(ch)
			}
		}
		if b.Len() == 0 {
			continue
		}
		value, err := strconv.Atoi(b.String())
		if err != nil {
			return 0, fmt.Errorf("columns %d-%d: %w", start, end, err)
		}
		numbers = append(numbers, value)
	}

	if len(numbers) == 0 {
		return 0, fmt.Errorf("no numbers found in columns %d-%d", start, end)
	}

	switch op {
	case '+':
		sum := 0
		for _, n := range numbers {
			sum += n
		}
		return sum, nil
	case '*':
		product := 1
		for _, n := range numbers {
			product *= n
		}
		return product, nil
	}

	return 0, fmt.Errorf("unhandled operator %q", op)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: contests/Aoc/2025/6/solution.py
import argparse
import sys
from typing import List


def read_lines() -> List[str]:
    return [line.rstrip("\n") for line in sys.stdin]


def pad_lines(lines: List[str]) -> List[str]:
    width = max((len(line) for line in lines), default=0)
    return [line.ljust(width) for line in lines]


def solve(grid: List[str], part2: bool) -> int:
    if not grid:
        return 0

    height = len(grid)
    width = len(grid[0])
    op_row = height - 1

    separator = [all(grid[r][c] == " " for r in range(height)) for c in range(width)]
    has_separator = any(separator)

    def eval_block(start: int, end: int) -> int:
        op = next((ch for ch in grid[op_row][start:end] if ch != " "), None)
        if op not in {"+", "*"}:
            raise ValueError(f"invalid operator in columns {start}-{end}")

        numbers: List[int] = []
        if part2:
            for c in range(end - 1, start - 1, -1):
                digits = [grid[r][c] for r in range(op_row) if grid[r][c].isdigit()]
                if digits:
                    numbers.append(int("".join(digits)))
        else:
            for r in range(op_row):
                segment = grid[r][start:end].strip()
                if segment:
                    numbers.append(int(segment))

        if not numbers:
            raise ValueError(f"no numbers found in columns {start}-{end}")

        if op == "+":
            return sum(numbers)
        product = 1
        for n in numbers:
            product *= n
        return product

    if has_separator:
        total = 0
        c = 0
        while c < width:
            while c < width and separator[c]:
                c += 1
            start = c
            while c < width and not separator[c]:
                c += 1
            if start < c:
                total += eval_block(start, c)
        return total

    # Fallback: operator proximity.
    ops = [(idx, ch) for idx, ch in enumerate(grid[op_row]) if ch in {"+", "*"}]
    if not ops:
        raise ValueError("no operators found")

    def nearest_op(col: int) -> int:
        best_idx = 0
        best_dist = abs(col - ops[0][0])
        for i in range(1, len(ops)):
            dist = abs(col - ops[i][0])
            if dist < best_dist or (dist == best_dist and ops[i][0] > ops[best_idx][0]):
                best_idx = i
                best_dist = dist
        return best_idx

    buckets: List[List[int]] = [[] for _ in ops]

    if part2:
        for c in range(width):
            digits = [grid[r][c] for r in range(op_row) if grid[r][c].isdigit()]
            if not digits:
                continue
            value = int("".join(digits))
            buckets[nearest_op(c)].append(value)
    else:
        for r in range(op_row):
            row = grid[r]
            c = 0
            while c < width:
                if not row[c].isdigit():
                    c += 1
                    continue
                start = c
                while c < width and row[c].isdigit():
                    c += 1
                value = int(row[start:c])
                center = (start + c - 1) // 2
                buckets[nearest_op(center)].append(value)

    total = 0
    for (col, op), nums in zip(ops, buckets):
        if not nums:
            raise ValueError(f"no numbers found for operator at column {col}")
        if op == "+":
            total += sum(nums)
        else:
            prod = 1
            for n in nums:
                prod *= n
            total += prod
    return total


def main() -> None:
    parser = argparse.ArgumentParser(description="Solve Trash Compactor worksheet.")
    parser.add_argument("-part2", action="store_true", dest="part2", help="use part 2 rules (columns right-to-left)")
    args = parser.parse_args()

    grid = pad_lines(read_lines())
    try:
        result = solve(grid, args.part2)
    except Exception as exc:  # pylint: disable=broad-except
        print(f"error: {exc}", file=sys.stderr)
        sys.exit(1)
    print(result)


if __name__ == "__main__":
    main()
